Mathlib/MeasureTheory/MeasurableSpace/CountablyGenerated.lean

```diff
@@ -99,6 +99,128 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
+
+section SeparatesPoints
+
+/-- We say that a measurable space separates points if for any two distinct points,
+there is a measurable set containing one but not the other. -/
+class SeparatesPoints (α : Type*) [m : MeasurableSpace α] : Prop where
+  separates : ∀ x y : α, (∀ s, MeasurableSet s → (x ∈ s ↔ y ∈ s)) → x = y
+
+theorem SeparatesPoints.mono {m m' : MeasurableSpace α} [hsep : @SeparatesPoints _ m] (h : m ≤ m') :
+    @SeparatesPoints _ m' := @SeparatesPoints.mk _ m' fun _ _ hxy ↦
+    @SeparatesPoints.separates _ m hsep _ _ fun _ hs ↦ hxy _ (h _ hs)
+
+instance (priority := 100) Subtype.separatesPoints [MeasurableSpace α] [h : SeparatesPoints α]
+    {s : Set α} : SeparatesPoints s :=
+  ⟨fun _ _ hxy ↦ Subtype.val_injective $ h.1 _ _ fun _ ht ↦ hxy _ $ measurable_subtype_coe ht⟩
+
+/-- If the measurable space generated by `S` separates points,
+then this is witnessed by sets in `S`. -/
+theorem separating_of_generateFrom (S : Set (Set α))
+    [h : @SeparatesPoints α (generateFrom S)] :
+    ∀ x y : α, (∀ s ∈ S, x ∈ s ↔ y ∈ s) → x = y := by
+  letI := generateFrom S
+  intros x y hxy
+  rw [← forall_generateFrom_mem_iff_mem_iff] at hxy
+  exact h.separates _ _ hxy
+
+instance (priority := 100) separatesPoints_of_measurableSingletonClass [MeasurableSpace α]
+    [MeasurableSingletonClass α] : SeparatesPoints α := by
+  refine' ⟨fun x y h ↦ _⟩
+  specialize h _ (MeasurableSet.singleton y)
+  simp_rw [mem_singleton_iff, iff_true] at h
+  assumption
+
+instance(priority := 100) [MeasurableSpace α] {s : Set α}
+    [h : CountablyGenerated s] [SeparatesPoints s] :
+    HasCountableSeparatingOn α MeasurableSet s := by
+  suffices HasCountableSeparatingOn s MeasurableSet univ from this.of_subtype fun _ ↦ id
+  rcases h.1 with ⟨b, hbc, hb⟩
+  refine ⟨⟨b, hbc, fun t ht ↦ hb.symm ▸ .basic t ht, ?_⟩⟩
+  rw [hb] at ‹SeparatesPoints s›
+  convert separating_of_generateFrom b
+  simp
+
+variable (α)
+
+/-- If a measurable space admits a countable separating family of measurable sets,
+there is a countably generated coarser space which still separates points. -/
+theorem exists_countablyGenerated_le_of_hasCountableSeparatingOn [m : MeasurableSpace α]
+    [h : HasCountableSeparatingOn α MeasurableSet univ] :
+    ∃ m' : MeasurableSpace α, @CountablyGenerated _ m' ∧ @SeparatesPoints _ m' ∧ m' ≤ m := by
+  rcases h.1 with ⟨b, bct, hbm, hb⟩
+  refine' ⟨generateFrom b, _, _, generateFrom_le hbm⟩
+  · use b
+  exact @SeparatesPoints.mk _ (generateFrom b) fun x y hxy ↦ hb _ trivial _ trivial
+    fun s hs ↦ hxy _ $ measurableSet_generateFrom hs
+
+open scoped Classical
+
+open Function
+
+/-- If a measurable space is countably generated and separates points, it is measure equivalent
+to some some subset of the Cantor space `ℕ → Bool` (equipped with the product sigma algebra). -/
+theorem measurableEquiv_nat_bool_of_countablyGenerated [MeasurableSpace α]
+    [CountablyGenerated α] [SeparatesPoints α] :
+    ∃ s : Set (ℕ → Bool), Nonempty (α ≃ᵐ s) := by
```

Comment:
What you are really doing here is construct a `MeasurableEmbedding` from `α` to `ℕ → Bool`. Could you add a definition for that embedding? (extract your construction of `f` from the proof).
Then that theorem is simply the statement that a measurable embedding is an equivalence to its range (and hence the theorem may not be needed).

Reply:
This is incorrect. There is no guarantee that the set `s` in this theorem will be measurable. In fact this will happen if and only if `\alpha` is standard Borel.

Reply:
I will add to the docstring to clarify this, though.

Reply:
You are right, sorry for the mistake.
Anyway, my main point is about the code style and not the math: in the proof you are constructing a function which is measurable and injective as a step towards building the equivalence and putting all that in an exists statement. In the next lemma, you are getting the equivalence by using choice and getting a measurable injective function from that.

The better alternative I am trying to describe is something like the following (with better names, and possibly with a different def for e that does not go through exists_seq_generateFrom):
```
noncomputable
def e [MeasurableSpace α] [CountablyGenerated α] [SeparatesPoints α] :=
  (exists_seq_generateFrom α).choose

lemma generateFrom_range_e [m : MeasurableSpace α] [CountablyGenerated α] [SeparatesPoints α] :
  m = generateFrom (range (e α)) := (exists_seq_generateFrom α).choose_spec

noncomputable
def f [MeasurableSpace α] [CountablyGenerated α] [SeparatesPoints α]
    (a : α) (n : ℕ) : Bool := a ∈ e α n

lemma injective_f [MeasurableSpace α] [CountablyGenerated α] [h : SeparatesPoints α] :
    Injective (f α) := by
  intro x y hxy
  rw [generateFrom_range_e α] at h
  apply separating_of_generateFrom (range (e α))
  rintro - ⟨n, rfl⟩
  rw [← decide_eq_decide]
  exact congr_fun hxy n

lemma measurable_f [MeasurableSpace α] [CountablyGenerated α] [SeparatesPoints α] :
    Measurable (f α) := by
  rw [measurable_pi_iff]
  refine fun n ↦ measurable_to_bool ?_
  simp only [preimage, mem_singleton_iff, f, Bool.decide_iff, setOf_mem_eq]
  simp_rw [generateFrom_range_e α]
  exact measurableSet_generateFrom $ mem_range_self _

/-- If a measurable space is countably generated and separates points, it is measure equivalent
to some some subset of the Cantor space `ℕ → Bool` (equipped with the product sigma algebra). -/
theorem measurableEquiv_nat_bool_of_countablyGenerated [MeasurableSpace α]
    [CountablyGenerated α] [SeparatesPoints α] :
    ∃ s : Set (ℕ → Bool), Nonempty (α ≃ᵐ s) := by
  use range (f α), Equiv.ofInjective _ (injective_f α)
  · apply Measurable.subtype_mk
    exact measurable_f α
  simp_rw [generateFrom_range_e α]
  apply measurable_generateFrom
  rintro _ ⟨n, rfl⟩
  rw [← Equiv.image_eq_preimage _ _]
  refine' ⟨{y | y n}, by measurability, _⟩
  rw [← Equiv.preimage_eq_iff_eq_image]
  simp[f]

/-- If a measurable space admits a countable sequence of measurable sets separating points,
it admits a measurable injection into the Cantor space `ℕ → Bool`
(equipped with the product sigma algebra). -/
theorem measurable_injection_nat_bool_of_hasCountableSeparatingOn [MeasurableSpace α]
    [HasCountableSeparatingOn α MeasurableSet univ] :
    ∃ f : α → ℕ → Bool, Measurable f ∧ Injective f := by
  rcases exists_countablyGenerated_le_of_hasCountableSeparatingOn α with ⟨m', _, _, m'le⟩
  refine' ⟨f α, _, injective_f α⟩
  exact (measurable_f α).mono m'le (by rfl)
```

Reply:
I agree with making `e` a definition; `exists_seq_generateFrom` was from awhile ago, before `countableGeneratingSet` existed. 

`f`, on the other hand, seems pretty specific to this proof. Perhaps it and the lemmas about it should be `private`?